cpp/src/parquet/encoding.cc

```diff
@@ -3033,12 +3079,243 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
+    std::string_view last_value_view = last_value_;
+
+    PARQUET_THROW_NOT_OK(::arrow::VisitArraySpanInline<typename ArrayType::TypeClass>(
+        *array.data(),
+        [&](::std::string_view view) {
+          if (ARROW_PREDICT_FALSE(view.size() >= kMaxByteArraySize)) {
+            return Status::Invalid("Parquet cannot store strings with size 2GB or more");
+          }
+          // Convert to ByteArray, so it can be passed to the suffix_encoder_.
+          const ByteArray src{view};
+
+          uint32_t j = 0;
+          const uint32_t common_length = std::min(previous_len, src.len);
+          while (j < common_length) {
+            if (last_value_view[j] != view[j]) {
+              break;
+            }
+            j++;
+          }
+          previous_len = src.len;
+          prefix_length_encoder_.Put({static_cast<int32_t>(j)}, 1);
+
+          last_value_view = view;
+          const auto suffix_length = static_cast<uint32_t>(src.len - j);
+          if (suffix_length == 0) {
+            suffix_encoder_.Put(&kEmpty, 1);
+            return Status::OK();
+          }
+          const uint8_t* suffix_ptr = src.ptr + j;
+          // Convert to ByteArray, so it can be passed to the suffix_encoder_.
+          const ByteArray suffix(suffix_length, suffix_ptr);
+          suffix_encoder_.Put(&suffix, 1);
+
+          return Status::OK();
+        },
+        []() { return Status::OK(); }));
+    last_value_ = last_value_view;
+  }
+
+  ::arrow::BufferBuilder sink_;
+  DeltaBitPackEncoder<Int32Type> prefix_length_encoder_;
+  DeltaLengthByteArrayEncoder<ByteArrayType> suffix_encoder_;
+  std::string last_value_;
+  const ByteArray kEmpty;
+  ArrowPoolVector<int32_t> prefix_lengths_;
+  static constexpr int kBatchSize_ = 256;
+};
+
+template <>
+void DeltaByteArrayEncoder<ByteArrayType>::Put(const ByteArray* src, int num_values) {
+  if (num_values == 0) {
+    return;
+  }
+
+  std::string_view last_value_view = last_value_;
+
+  for (int i = 0; i < num_values; i += kBatchSize_) {
+    const int batch_size = std::min(kBatchSize_, num_values - i);
+
+    for (int j = 0; j < batch_size; ++j) {
+      // Convert to ByteArray, so we can pass to the suffix_encoder_.
+      const ByteArray value = src[i + j];
+      if (ARROW_PREDICT_FALSE(value.len >= static_cast<int32_t>(kMaxByteArraySize))) {
+        throw ParquetException("Parquet cannot store strings with size 2GB or more");
+      }
+      auto view = std::string_view{value};
+
+      uint32_t k = 0;
+      const uint32_t common_length =
+          std::min(value.len, static_cast<uint32_t>(last_value_view.length()));
+      while (k < common_length) {
+        if (last_value_view[k] != view[k]) {
+          break;
+        }
+        k++;
+      }
+
+      last_value_view = view;
+      prefix_lengths_[j] = k;
+      const auto suffix_length = static_cast<uint32_t>(value.len - k);
+
+      if (suffix_length == 0) {
+        continue;
+      }
```

Comment:
I don't understand. You still need to write a 0-size string to the suffix encoder.
```suggestion
```